editor/plugins/gizmos/particles_3d_emission_shape_gizmo_plugin.cpp

```diff
@@ -0,0 +1,748 @@
+				ur->commit_action();
+			}
+		} else if (shape == CPUParticles3D::EMISSION_SHAPE_RING) {
+			if (p_id == 0) {
+				if (p_cancel) {
+					particles->set_emission_ring_axis(p_restore);
+				} else {
+					EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
+					ur->create_action(TTR("Change emission ring axis"));
+					ur->add_do_property(particles, "emission_ring_axis", particles->get_emission_ring_axis());
+					ur->add_undo_property(particles, "emission_ring_axis", p_restore);
+					ur->commit_action();
+				}
+			} else if (p_id == 1) {
+				if (p_cancel) {
+					particles->set_emission_ring_height(p_restore);
+				} else {
+					EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
+					ur->create_action(TTR("Change emission ring height"));
+					ur->add_do_property(particles, "emission_ring_height", particles->get_emission_ring_height());
+					ur->add_undo_property(particles, "emission_ring_height", p_restore);
+					ur->commit_action();
+				}
+			} else if (p_id == 2) {
+				if (p_cancel) {
+					particles->set_emission_ring_radius(p_restore);
+				} else {
+					EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
+					ur->create_action(TTR("Change emission ring radius"));
+					ur->add_do_property(particles, "emission_ring_radius", particles->get_emission_ring_radius());
+					ur->add_undo_property(particles, "emission_ring_radius", p_restore);
+					ur->commit_action();
+				}
+			} else {
+				if (p_cancel) {
+					particles->set_emission_ring_inner_radius(p_restore);
+				} else {
+					EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
+					ur->create_action(TTR("Change emission ring inner radius"));
+					ur->add_do_property(particles, "emission_ring_inner_radius", particles->get_emission_ring_inner_radius());
+					ur->add_undo_property(particles, "emission_ring_inner_radius", p_restore);
+					ur->commit_action();
+				}
+			}
+		}
+	}
+}
+
+void Particles3DEmissionShapeGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
+	p_gizmo->clear();
+
+	if (Object::cast_to<GPUParticles3D>(p_gizmo->get_node_3d())) {
+		GPUParticles3D *particles = Object::cast_to<GPUParticles3D>(p_gizmo->get_node_3d());
+
+		if (particles->get_process_material() != nullptr) {
+			Ref<ParticleProcessMaterial> mat = particles->get_process_material();
+			ParticleProcessMaterial::EmissionShape shape = mat->get_emission_shape();
+
+			const Ref<Material> material = get_material("particles_emission_shape_material", p_gizmo);
+			Ref<Material> handles_material = get_material("handles");
+
+			if (shape == ParticleProcessMaterial::EMISSION_SHAPE_SPHERE || shape == ParticleProcessMaterial::EMISSION_SHAPE_SPHERE_SURFACE) {
+				Vector3 offset = mat->get_emission_shape_offset();
+				Vector3 scale = mat->get_emission_shape_scale();
+
+				float r = mat->get_emission_sphere_radius();
+				Vector<Vector3> points;
+				for (int i = 0; i <= 360; i++) {
```

Comment:
I'd decrease the detail level to 120 like in OmniLight3D/SpotLight3D. 360 can be quite slow on integrated graphics, particularly with multiple nodes selected.

Reply:
These suggestions are still unresolved, WDYT @paddy-exe?

Reply:
May I add, I would put this value in a constant to make the intent clearer.

Reply:
Sorry I have just seen the comments. I will add those changes. They seem plausible.

Reply:
Adjusted now as requested